File: flow/actorcompiler/ParseTree.cs
/*
 * ParseTree.cs
 *
 * This source file is part of the FoundationDB open source project
 *
 * Copyright 2013-2024 Apple Inc. and the FoundationDB project authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Text.RegularExpressions;

namespace actorcompiler
{
    class VarDeclaration
    {
        public string type;
        public string name;
        public string initializer;
        public bool initializerConstructorSyntax;
    };

    abstract class Statement
    {
        public int FirstSourceLine;

        public virtual bool containsWait()
        {
            return false;
        }
    };

    class PlainOldCodeStatement : Statement
    {
        public string code;

        public override string ToString()
        {
            return code;
        }
    };

    class StateDeclarationStatement : Statement
    {
        public VarDeclaration decl;

        public override string ToString()
        {
            if (decl.initializerConstructorSyntax)
                return string.Format("State {0} {1}({2});", decl.type, decl.name, decl.initializer);
            else
                return string.Format(
                    "State {0} {1} = {2};",
                    decl.type,
                    decl.name,
                    decl.initializer
                );
        }
    };

    class WhileStatement : Statement
    {
        public string expression;
        public Statement body;

        public override bool containsWait()
        {
            return body.containsWait();
        }
    };

    class ForStatement : Statement
    {
        public string initExpression = "";
        public string condExpression = "";
        public string nextExpression = "";
        public Statement body;

        public override bool containsWait()
        {
            return body.containsWait();
        }
    };

    class RangeForStatement : Statement
    {
        public string rangeExpression;
        public string rangeDecl;
        public Statement body;

        public override bool containsWait()
        {
            return body.containsWait();
        }
    };

    class LoopStatement : Statement
    {
        public Statement body;

        public override string ToString()
        {
            return "Loop " + body.ToString();
        }

        public override bool containsWait()
        {
            return body.containsWait();
        }
    };

    class BreakStatement : Statement { };

    class ContinueStatement : Statement { };

    class IfStatement : Statement
    {
        public string expression;
        public bool constexpr;
        public Statement ifBody;
        public Statement elseBody; // might be null

        public override bool containsWait()
        {
            return ifBody.containsWait() || (elseBody != null && elseBody.containsWait());
        }
    };

    class ReturnStatement : Statement
    {
        public string expression;

        public override string ToString()
        {
            return "Return " + expression;
        }
    };

    class WaitStatement : Statement
    {
        public VarDeclaration result;
        public string futureExpression;
        public bool resultIsState;
        public bool isWaitNext;

        public override string ToString()
        {
            return string.Format(
                "Wait {0} {1} <- {2} ({3})",
                result.type,
                result.name,
                futureExpression,
                resultIsState ? "state" : "local"
            );
        }

        public override bool containsWait()
        {
            return true;
        }
    };

    class ChooseStatement : Statement
    {
        public Statement body;

        public override string ToString()
        {
            return "Choose " + body.ToString();
        }

        public override bool containsWait()
        {
            return body.containsWait();
        }
    };

    class WhenStatement : Statement
    {
        public WaitStatement wait;
        public Statement body;

        public override string ToString()
        {
            return string.Format("When ({0}) {1}", wait, body);
        }

        public override bool containsWait()
        {
            return true;
        }
    };

    class TryStatement : Statement
    {
        public struct Catch
        {
            public string expression;
            public Statement body;
            public int FirstSourceLine;
        };

        public Statement tryBody;
        public List<Catch> catches;

        public override bool containsWait()
        {
            if (tryBody.containsWait())
                return true;
            foreach (Catch c in catches)
                if (c.body.containsWait())
                    return true;
            return false;
        }
    };

    class ThrowStatement : Statement
    {
        public string expression;
    };

    class CodeBlock : Statement
    {
        public Statement[] statements;

        public override string ToString()
        {
            return string.Join(
                "\n",
                new string[] { "CodeBlock" }
                    .Concat(statements.Select(s => s.ToString()))
                    .Concat(new string[] { "EndCodeBlock" })
                    .ToArray()
            );
        }

        public override bool containsWait()
        {
            foreach (Statement s in statements)
                if (s.containsWait())
                    return true;
            return false;
        }
    };

    class Declaration
    {
        public string type;
        public string name;
        public string comment;
    };

    class Actor
    {
        public List<string> attributes = new List<string>();
        public string returnType;
        public string name;
        public string enclosingClass = null;
        public VarDeclaration[] parameters;
        public VarDeclaration[] templateFormals; //< null if not a template
        public CodeBlock body;
        public int SourceLine;
        public bool isStatic = false;
        private bool isUncancellable;
        public string testCaseParameters = null;
        public string nameSpace = null;
        public bool isForwardDeclaration = false;
        public bool isTestCase = false;

        public bool IsCancellable()
        {
            return returnType != null && !isUncancellable;
        }

        public void SetUncancellable()
        {
            isUncancellable = true;
        }
    };

    class Descr
    {
        public string name;
        public string superClassList;
        public List<Declaration> body;
    };
};

File: flow/actorcompiler/parse_tree.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import List, Optional, Sequence


@dataclass
class VarDeclaration:
    type: str = ""
    name: str = ""
    initializer: Optional[str] = ""
    initializerConstructorSyntax: bool = False


@dataclass
class Statement:
    FirstSourceLine: int = 0

    def containsWait(self) -> bool:
        return False


@dataclass
class PlainOldCodeStatement(Statement):
    code: str = ""

    def __str__(self) -> str:
        return self.code


@dataclass
class StateDeclarationStatement(Statement):
    decl: VarDeclaration = field(default_factory=VarDeclaration)

    def __str__(self) -> str:
        if self.decl.initializerConstructorSyntax:
            return f"State {self.decl.type} {self.decl.name}({self.decl.initializer});"
        return f"State {self.decl.type} {self.decl.name} = {self.decl.initializer};"


@dataclass
class WhileStatement(Statement):
    expression: str = ""
    body: Statement = field(default_factory=Statement)

    def containsWait(self) -> bool:
        return self.body.containsWait()


@dataclass
class ForStatement(Statement):
    initExpression: str = ""
    condExpression: str = ""
    nextExpression: str = ""
    body: Statement = field(default_factory=Statement)

    def containsWait(self) -> bool:
        return self.body.containsWait()


@dataclass
class RangeForStatement(Statement):
    rangeExpression: str = ""
    rangeDecl: str = ""
    body: Statement = field(default_factory=Statement)

    def containsWait(self) -> bool:
        return self.body.containsWait()


@dataclass
class LoopStatement(Statement):
    body: Statement = field(default_factory=Statement)

    def __str__(self) -> str:
        return f"Loop {self.body}"

    def containsWait(self) -> bool:
        return self.body.containsWait()


@dataclass
class BreakStatement(Statement):
    pass


@dataclass
class ContinueStatement(Statement):
    pass


@dataclass
class IfStatement(Statement):
    expression: str = ""
    constexpr: bool = False
    ifBody: Statement = field(default_factory=Statement)
    elseBody: Optional[Statement] = None

    def containsWait(self) -> bool:
        return self.ifBody.containsWait() or (
            self.elseBody is not None and self.elseBody.containsWait()
        )


@dataclass
class ReturnStatement(Statement):
    expression: str = ""

    def __str__(self) -> str:
        return f"Return {self.expression}"


@dataclass
class WaitStatement(Statement):
    result: VarDeclaration = field(default_factory=VarDeclaration)
    futureExpression: str = ""
    resultIsState: bool = False
    isWaitNext: bool = False

    def __str__(self) -> str:
        return f"Wait {self.result.type} {self.result.name} <- {self.futureExpression} ({'state' if self.resultIsState else 'local'})"

    def containsWait(self) -> bool:
        return True


@dataclass
class ChooseStatement(Statement):
    body: Statement = field(default_factory=Statement)

    def __str__(self) -> str:
        return f"Choose {self.body}"

    def containsWait(self) -> bool:
        return self.body.containsWait()


@dataclass
class WhenStatement(Statement):
    wait: WaitStatement = field(default_factory=WaitStatement)
    body: Optional[Statement] = None

    def __str__(self) -> str:
        return f"When ({self.wait}) {self.body}"

    def containsWait(self) -> bool:
        return True


@dataclass
class TryStatement(Statement):
    @dataclass
    class Catch:
        expression: str = ""
        body: Statement = field(default_factory=Statement)
        FirstSourceLine: int = 0

    tryBody: Statement = field(default_factory=Statement)
    catches: List["TryStatement.Catch"] = field(default_factory=list)

    def containsWait(self) -> bool:
        if self.tryBody.containsWait():
            return True
        return any(c.body.containsWait() for c in self.catches)


@dataclass
class ThrowStatement(Statement):
    expression: str = ""


@dataclass
class CodeBlock(Statement):
    statements: Sequence[Statement] = field(default_factory=list)

    def __str__(self) -> str:
        joined = "\n".join(str(stmt) for stmt in self.statements)
        return f"CodeBlock\n{joined}\nEndCodeBlock"

    def containsWait(self) -> bool:
        return any(stmt.containsWait() for stmt in self.statements)


@dataclass
class Declaration:
    type: str = ""
    name: str = ""
    comment: str = ""


@dataclass
class Actor:
    attributes: List[str] = field(default_factory=list)
    returnType: Optional[str] = None
    name: str = ""
    enclosingClass: Optional[str] = None
    parameters: Sequence[VarDeclaration] = field(default_factory=list)
    templateFormals: Optional[Sequence[VarDeclaration]] = None
    body: CodeBlock = field(default_factory=CodeBlock)
    SourceLine: int = 0
    isStatic: bool = False
    _isUncancellable: bool = False
    testCaseParameters: Optional[str] = None
    nameSpace: Optional[str] = None
    isForwardDeclaration: bool = False
    isTestCase: bool = False

    def IsCancellable(self) -> bool:
        return self.returnType is not None and not self._isUncancellable

    def SetUncancellable(self) -> None:
        self._isUncancellable = True


@dataclass
class Descr:
    name: str = ""
    superClassList: Optional[str] = None
    body: List[Declaration] = field(default_factory=list)
